### ISSUE TITLE
Missing package in packrat restore

### ISSUE THREAD

**Issue by mmgledhill** [rstudio/rsconnect#351]
We'd like to use packrat to isolate/leverage a github version of plotly. We confirmed the github package details are in the packrat.lock file. When the app is deployed via RStudio/rsconnect, it writes a packrat restore to the app logs. The package is not included in the package restore and the app ends up downloading the latest version from CRAN when it first runs. I'm pretty new to packrat. Is there something else (an option or something) that we need to do to force rsconnect to use our packrat.lock file and/or github version? The package was just updated in CRAN.

As fyi:
 packageVersion("rsconnect")
[1] ‘0.8.13’

From packrat.lock:

PackratFormat: 1.4
PackratVersion: 0.5.0
RVersion: 3.5.2
Repos: CRAN=https://cran.cnr.berkeley.edu/,
    we_cran=http://some.custom.my.cran.com,
    my_cran=http://another.custom.cran.com/

< more packages >

Package: plotly
Source: github
**Version: 4.8.0.9000**
Hash: 883cd71cbb4a0bf241b60ff373981bbf
Requires: RColorBrewer, base64enc, crosstalk, data.table, digest,
    dplyr, ggplot2, hexbin, htmltools, htmlwidgets, httr, jsonlite,
    lazyeval, magrittr, promises, purrr, rlang, scales, tibble, tidyr,
    viridisLite
GithubRepo: plotly
GithubUsername: ropensci
GithubRef: master
GithubSha1: 8a0470d7e406b9bbd4542f111952da780b0672e0

From app log:

2019/04/10 14:33:20.203381902 
2019/04/10 14:33:20.212633192 trying URL '**https://cran.cnr.berkeley.edu/src/contrib/plotly_4.9.0.tar.gz**'
2019/04/10 14:33:20.543796664 Content type 'application/x-gzip' length 3469680 bytes (3.3 MB)
2019/04/10 14:33:20.920050893 ==================================================
2019/04/10 14:33:20.920077431 downloaded 3.3 MB
2019/04/10 14:33:20.920127567

**Comment by kevinushey** [comment]
Currently, `rsconnect` ignores any pre-existing Packrat lockfile and instead generates its own on demand. I think this implies that if you were attempting to deploy your application, if you had plotly 4.9.0 (from CRAN) installed in your R library, that version would be used instead.

Unfortunately, I don't think there is a mechanism to use a previously-existing Packrat lockfile. Would you mind re-filing this issue at https://github.com/rstudio/rsconnect/issues? I think any requisite change would need to happen there.

**Comment by mmgledhill** [comment]
Thanks for the additional info.

Does it use my User Library or the global library on the server? I removed my CRAN version from my User Library and reinstalled from github but I am still seeing the issue after restarting Rstudio and redeploying. Package is still missing from Packrat restore, and it is still pulling down wrong version from CRAN. I think this older issue  #227  may be related somehow....at least we are both trying to figure out how to use a packrat version of a package. I can resubmit another issue (unless there is a way to rename this one) but I thought I was already under https://github.com/rstudio/rsconnect/issues.

I tried to reference this [documentation in section 17.3.5](https://docs.rstudio.com/connect/1.4.4/admin/appendix-deployment-guide.html#step-3-bundle-is-deployed-on-connect) and rsconnect:::performPackratSnapshot(appDir) to check the packrat.lock file that rsconnect uses but I don't think it is writing out the actual file since it includes the older github version that is in my Packrat library and not my User library (which is a newer version from github now)....unless it is writing this file somewhere other than my ../appdir/packrat folder.

When packrat is initiated my .libPaths are based on my project but it sounds like it is using a different libPath on rsconnect deployment? Documentation suggests that using packrat should be the cure/fix to use a different package version on server but it's not using the packrat.lock file or it's creating another one that I can't access on my side?

Proof of me uninstalling/reinstalling plotly to User library pre redeployment above:
`> .libPaths()
[1] "**C:/Users/mmgledhill/Documents/R/win-library/3.5**" "C:/Program Files/R/R-3.5.2/library"             
> remove.packages("plotly")
Removing package from ‘C:/Users/mmgledhill/Documents/R/win-library/3.5’
(as ‘lib’ is unspecified)
> devtools::install_github("ropensci/plotly")
Downloading GitHub repo ropensci/plotly@master
from URL https://api.github.com/repos/ropensci/plotly/zipball/master
Installing plotly
"C:/PROGRA~1/R/R-35~1.2/bin/x64/R" --no-site-file --no-environ --no-save --no-restore --quiet CMD INSTALL  \
  "C:/Users/mmgledhill/AppData/Local/Temp/RtmpCCerVN/devtools1847c56786763/ropensci-plotly-4763648"  \
  --library="**C:/Users/mmgledhill/Documents/R/win-library/3.5**" --install-tests 
Loading required package: devtools
\* installing *source* package 'plotly' ...`

**Comment by aronatkins** [comment]
It looks like you are using the RStudio IDE with `rsconnect` to deploy content to RStudio Connect. Could you let us know what RStudio IDE version you are using? There have occasionally been problems where the IDE would not propagate `.libPaths` across its different R contexts. @kevinushey would be able to tell us if your IDE version may experience one of these issues.

The `rsconnect` package uses `packrat` to compute the dependencies for your content and does that by analyzing the package versions available in the current R environment. It assumes that any initialization of your development environment by `packrat` has already occurred. Since `rsconnect`+`packrat` inspects the R environment, that should mean packages are discovered within your packrat project directory.

Your development `packrat` directory and `packrat.lock` files are ignored when computing a snapshot prior to deployment. The bundle deployed to Connect contains a lockfile computed from the environment analysis.

The lockfile that lands in the bundle is _not_ the `packrat/packrat.lock` file in your local directory.

There are a couple ways you can inspect the final result:

### Download the bundle archive file (`.tar.gz`) from RStudio Connect.

The "Source Versions" menu item gives you information about each bundle you have deployed against a specific piece of content. Connect retains the archive file and lets you download it. This is the exact `.tar.gz` created by `rsconnect` and uploaded on your behalf.

A bundle is created and uploaded even when its packages cannot be fully installed/restored.

You may need to filter by "Content You Own" to see that content in Connect; the default visibility filter does not show incomplete content.

### Create a bundle yourself with `rsconnect:::bundleApp`.

This will create a `.tar.gz` file in a temporary directory. You'll need to supply all its arguments (which are normally computed by `rsconnect::deployApp`.

Here's an example where I am bundling one of my Shiny applications. I am providing all the arguments to `:::bundleApp` so you can see them, but many are given their default values.

```r
rsconnect:::bundleApp(appName = "simple", appDir = ".", 
      appFiles = c("app.R"), appPrimaryDoc = NULL, 
      assetTypeName = "application", contentCategory = NULL, 
      verbose = FALSE, python = NULL)
# => [1] "/var/folders/ts/s940qvdj5vj1czr9qh07fvtw0000gn/T//Rtmp7mVjHX/rsconnect-bundlef69c7326da4b.tar.gz"
```

Using `tar` from the terminal, we can list the included files:

```bash
tar ztf /var/folders/ts/s940qvdj5vj1czr9qh07fvtw0000gn/T//Rtmp7mVjHX/rsconnect-bundlef69c7326da4b.tar.gz | grep -v packrat/desc
./app.R
./index.htm
./manifest.json
./packrat/
./packrat/packrat.lock
```

**Comment by aronatkins** [comment]
One word of caution: That `tar.gz` file is a flat archive and introduces no directory hierarchy. Do not un-tar that archive in your source directory; you'll overwrite your files.

**Comment by mmgledhill** [comment]
This is exactly what I was looking for! Who needs rsconnect:::performPackratSnapshot(appDir) when you can just download and untar to see the manifest.json and packrat.lock. ;)

Plotly package is missing from the manifest.json AND packrat.lock file in the bundle/tar that is deployed. Does it have something to do with me using a github repo? It is loaded at same time as other packages with xfun::pkg_attach2. 

Here is my Rstudio info:
versionInfo()$version
[1] ‘1.1.463’

**Comment by aronatkins** [comment]
It's the call to `xfun::pkg_attach2`. When `packrat` scans your code, it tries to identify the dependencies in play. It isn't executing your code, so it relies on some heuristics.

https://github.com/rstudio/packrat/blob/1ff9fbd6b1874c0332c8b94b7c846c9a0470c339/R/dependencies.R#L440-L443

`packrat` doesn't scan for `pkg_attach2`.

There are other techniques in play (looking for `libname::funcname`, etc), but the missed library scanning feels likely.

@kevinushey - thoughts?

@mmgledhill - could you try explicit `library` statements as a workaround?

**Comment by mmgledhill** [comment]
That fixed it. Once there is explicit library call for plotly, it showed up in the manifest.json with the github source.  It is curious that when I snap a local packrat snapshot it is included in my local packrat.lock even though I'm using the same xfun pkg_attach2 call? Is there something different about how rsconnect does it vs. a direct `packrat::snapshot()` call?

Lessons learned:

- rsconnect does not use local packrat.lock but creates it's own. Make sure whatever version you need is installed in your local .libPaths() location.
- To verify manifest.json and packrat.lock for app that is deployed, download from server and `untar` in separate folder.
- Packages attached with xfun `pkg_attach2` will not be included in manifest.json or packrat.lock when deployed to server. You must use explicit `library`, `require`, `loadNamespace` or `requireNamespace` call.

**Comment by cderv** [comment]
> rsconnect does not use local packrat.lock but creates it's own. Make sure whatever version you need is installed in your local .libPaths() location.

Just to share experience about this, What I do is use packrat in my project to isolate versions. Then I work inside this packrat project (with `packrat::on()` ). When I want to deploy, I stay inside that project that use packrat and `rsconnect` will find the library I used by lookin for version inside the local packrat project library because `packrat::on()` has modified my `.libPaths()`. This allows me to easily deploy different projects with different versions. 

> To verify manifest.json and packrat.lock for app that is deployed, download from server and untar in separate folder.

`rsconnect` has also now a `rsconnect::writeManifest()` that generate the manifest.json to commit locally in the git project. It allows for some (experimental for now) API deployment [using RStudio Connect for example](https://github.com/rstudio/connect-api-deploy-shiny). It may help somehow.

**Comment by kevinushey** [comment]
Packrat's snapshot effectively tries to take the union of:

1. What is sees in your active R library (as reported by `.libPaths()`);
2. What it discovers through dependency discovery in your project (e.g. by looking for `library()` calls). (In such a case, the latest version of the package as available on CRAN is used)

This implies that if you're working with Packrat but want to deploy using the same package versions as in the private library, you need to activate that project first (as described by @cderv).
